# Kubernetes Storage

Two ways of storing data:
- Volumes attached to Pod: Lost if pods are destroyed, but used to exchange data between containers on one and the same pod
- Mounted external storage: Persistent storage

### 1. Volumes to share data between Containers

Ways to store transient data:
- **emptyDir**: Simple folder to store transient data (e.g. in memory)
- **hostPath**: Store data on node level (still transient)

### 2. Persistent Storage

Two-Step Process of Mounting/ Using persistent Storage (local file storage, cloud storage)
1. The cluster administrator sets up the underlying storage and registers it in Kubernetes by creating a **PersistentVolume** resource. When creating the PersistentVolume, the admin specifies its size and the access modes it supports.
2. When a cluster user needs to use persistent storage in one of their pods, they first create a **PersistentVolumeClaim** manifest, specifying the minimum size and the access mode they require. The user then submits the PersistentVolumeClaim manifest to the Kubernetes API server, and Kubernetes finds the appropriate PersistentVolume and binds the volume to the claim. The PersistentVolumeClaim can then be used as one of the volumes inside a pod

**Example: MongoDB with local persistent storage**

**First step**: Create local persistent volume and volume claim

In [29]:
%%bash
kubectl apply -f persistent_volume.yaml

persistentvolume/local-volume unchanged


In [30]:
%%bash
kubectl apply -f persistent_volume_claim.yaml

persistentvolumeclaim/local-volume-claim unchanged


In [31]:
%%bash
kubectl get pv

NAME           CAPACITY   ACCESS MODES   RECLAIM POLICY   STATUS   CLAIM                        STORAGECLASS   REASON   AGE
local-volume   1Gi        RWO            Delete           Bound    default/local-volume-claim   manual                  52m


In [32]:
%%bash
kubectl get pvc

NAME                 STATUS   VOLUME         CAPACITY   ACCESS MODES   STORAGECLASS   AGE
local-volume-claim   Bound    local-volume   1Gi        RWO            manual         52m


**Second Step**: Add Persistent Volume Claim to Pod: mongodb_pod.yaml

In [46]:
%%bash
kubectl apply -f mongodb_pod.yaml

pod/mongodb-pod created


In [47]:
%%bash
kubectl get pods --show-labels

NAME          READY   STATUS              RESTARTS   AGE   LABELS
mongodb-pod   0/1     ContainerCreating   0          1s    app=mongodb


**Third Step**: Add LoadBalancer Service to access MongoDB Pod: mongodb_svc.yaml

In [48]:
%%bash
kubectl apply -f mongodb_svc.yaml

service/mongodb-service created


In [49]:
%%bash
kubectl get services

NAME              TYPE           CLUSTER-IP       EXTERNAL-IP   PORT(S)           AGE
kubernetes        ClusterIP      10.96.0.1        <none>        443/TCP           9s
mongodb-service   LoadBalancer   10.101.145.166   localhost     27017:31169/TCP   1s


**Access mongodb from Mongo Compass**: mongodb://localhost:27017

Access mongo container... in terminal

In [52]:
%%bash
kubectl exec -it mongodb-pod mongo

MongoDB shell version v4.2.8
connecting to: mongodb://127.0.0.1:27017/?compressors=disabled&gssapiServiceName=mongodb
Implicit session: session { "id" : UUID("0805a1a5-bf32-48bb-ab6c-c44e152ba607") }
MongoDB server version: 4.2.8
bye


Unable to use a TTY - input is not a terminal or the right kind of file


**In mongo shell create admin user**:
- use admin
- db.createUser(
  {
    user: "admin",
    pwd: "pw123",
    roles: [ { role: "userAdminAnyDatabase", db: "admin" } ]
  }

Now access from Mongo Compass: mongodb://admin:pw123@localhost:27017

**Now check out /tmp/mongodata on mac**

In [53]:
%%bash
kubectl delete all --all

pod "mongodb-pod" deleted
service "kubernetes" deleted
service "mongodb-service" deleted
